pkg/components/prometheus-operator/component_conversion_test.go

```diff
@@ -0,0 +1,171 @@
+package prometheus //nolint:testpackage
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+)
+
+func renderManifests(t *testing.T, configHCL string) map[string]string {
+	name := "prometheus-operator"
+
+	component := newComponent()
+
+	body, diagnostics := util.GetComponentBody(configHCL, name)
+	if diagnostics != nil {
+		t.Fatalf("Getting component body: %v", diagnostics)
+	}
+
+	diagnostics = component.LoadConfig(body, &hcl.EvalContext{})
+	if diagnostics.HasErrors() {
+		t.Fatalf("Valid config should not return an error, got: %v", diagnostics)
+	}
+
+	ret, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	return ret
+}
+
+//nolint:funlen
+func TestConversion(t *testing.T) {
+	testCases := []struct {
+		Name                   string
+		InputConfig            string
+		ExpectedConfigFileName string
+		Expected               reflect.Value
+		JSONPath               string
+	}{
+		{
+			Name: "use external_url param",
+			InputConfig: `
+component "prometheus-operator" {
+  prometheus {
+    external_url = "https://prometheus.externalurl.net"
+  }
+}
+`,
+			ExpectedConfigFileName: "prometheus-operator/templates/prometheus/prometheus.yaml",
+			Expected:               reflect.ValueOf("https://prometheus.externalurl.net"),
+			JSONPath:               "{.spec.externalUrl}",
+		},
+		{
+			Name: "no external_url param",
+			InputConfig: `
+		component "prometheus-operator" {
+		  prometheus {
+		    ingress {
+		      host                       = "prometheus.mydomain.net"
+		      class                      = "contour"
+		      certmanager_cluster_issuer = "letsencrypt-production"
+		    }
+		  }
+		}
+		`,
+			ExpectedConfigFileName: "prometheus-operator/templates/prometheus/prometheus.yaml",
+			Expected:               reflect.ValueOf("https://prometheus.mydomain.net"),
```

Comment:
Perhaps we could have `ExpectedString` field instead of require using `reflect.ValueOf` here? That would simplify writing tests and prevent putting erronous values here.

Or maybe `ExpectedStringValue`? To be consistent with `ExpectedConfigFileName` ?

Reply:
This was done to keep the test case struct generic for strings, bool, int, float, arrays, etc. We can check anything. And we don't have to write the test code multiple times.

Reply:
Yes, but I think with `ExpectedStringValue` we limit the API surface to what is actually supported. If I understand the code, if I use `reflect.ValueOf(false)`, the test will always fail.

Reply:
Yes it will fail. But it is incumbent on the test author to add support of that type. We can do it right now but it might be wasted efforts to do it if certain data type is not used. Also that's why we have `//nolint:exhaustive` in place.

Reply:
Right. But with `ExpectedStringValue` we have an testing API with clear expectations I think, hence my suggestion.

Reply:
So overtime there will be multiple `Expected<Type>Value` fields. Every time a new field type needs tested a new var needs adding in the struct?

Reply:
Yes, but they would all be optional (perhaps pointers to be able to also match zero values). I don't think we will need many types anyway. String, bool, int should be mostly sufficient I guess.

Reply:
So why not add them(more cases) to the type switch instead of adding such fields?

Reply:
Because switch is an implementation and field is your API. I don't care much about the implementation, but I would like the API for writing tests to be clear and robust, so we don't spend time figuring out why things are not working.

Reply:
Yes, we can keep the API generic by doing the heavy lifting in the implementation.

Reply:
What do you mean? What I mean is, API capabilities should be aligned with the implementation. We currently support **all** types of values on the API side, but implementation only supports **string** type. This is a problem in my opinion.

Reply:
That's what I meant here: https://github.com/kinvolk/lokomotive/pull/964#discussion_r493384023. We can add support to the types as needed. Since this is not user facing thing code/test author can do it later.

If you insist we can do it now.

Reply:
No, I wouldn't add unused implementation and spend time to make it exhaustive, as I think most likely we won't need to support those types anyway. I'd make the API narrower instead.

Reply:
I'd say let's use `ExpectedString` and if in the future we need more types we can reevaluate.

Reply:
Alright, lets reevaluate when we add more..